File: app/routers/urls.py
from __future__ import annotations

import csv
import io
from typing import Annotated

from fastapi import APIRouter, File, Form, Request, UploadFile
from fastapi.responses import RedirectResponse, StreamingResponse
from sqlalchemy import func, select

from app.deps import CurrentUser, CurrentWorkspace, DbSession, flash, get_site_or_404
from app.models import IndexStatus, PageUrl, Site
from app.services import tasks
from app.services import urls as url_service
from app.services.scheduler import run_in_background
from app.services.sites import list_sites
from app.templating import render

router = APIRouter(prefix="/urls", tags=["urls"])

PER_PAGE = 100


@router.get("")
def urls_page(
    request: Request,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
    site_id: int = 0,
    status: str = "",
    q: str = "",
    page: int = 1,
):
    query = (
        select(PageUrl)
        .join(Site, Site.id == PageUrl.site_id)
        .where(Site.workspace_id == workspace.id)
    )
    count_query = (
        select(func.count(PageUrl.id))
        .join(Site, Site.id == PageUrl.site_id)
        .where(Site.workspace_id == workspace.id)
    )

    if site_id:
        query = query.where(PageUrl.site_id == site_id)
        count_query = count_query.where(PageUrl.site_id == site_id)
    if status:
        query = query.where(PageUrl.index_status == status)
        count_query = count_query.where(PageUrl.index_status == status)
    if q:
        query = query.where(PageUrl.url.contains(q))
        count_query = count_query.where(PageUrl.url.contains(q))

    total = db.scalar(count_query) or 0
    page = max(page, 1)
    rows = list(
        db.scalars(
            query.order_by(PageUrl.last_checked_at.is_(None).desc(), PageUrl.id.desc())
            .offset((page - 1) * PER_PAGE)
            .limit(PER_PAGE)
        )
    )
    site_map = {site.id: site for site in list_sites(db, workspace)}

    return render(
        request,
        "urls.html",
        {
            "user": user,
            "workspace": workspace,
            "rows": rows,
            "sites": list(site_map.values()),
            "site_map": site_map,
            "total": total,
            "page": page,
            "pages": max(1, (total + PER_PAGE - 1) // PER_PAGE),
            "filter_site": site_id,
            "filter_status": status,
            "query": q,
            "statuses": [s.value for s in IndexStatus],
            "stats": url_service.workspace_url_stats(db, workspace.id),
            "active_page": "urls",
        },
    )


@router.post("/add")
def add_urls(
    request: Request,
    workspace: CurrentWorkspace,
    db: DbSession,
    site_id: int = Form(...),
    urls_blob: str = Form(""),
    priority: int = Form(0),
    submit_now: bool = Form(False),
    file: Annotated[UploadFile | None, File()] = None,
):
    site = get_site_or_404(db, workspace, site_id)

    blob = urls_blob or ""
    if file is not None and file.filename:
        try:
            blob += "\n" + file.file.read().decode("utf-8", errors="ignore")
        except Exception:  # noqa: BLE001
            flash(request, "Nie udalo sie odczytac pliku.", "error")

    candidates = url_service.parse_url_blob(blob)
    if not candidates:
        flash(request, "Nie znaleziono zadnego poprawnego adresu URL.", "error")
        return RedirectResponse(f"/sites/{site_id}?tab=urls", status_code=303)

    from app.services.sitemap_parser import url_belongs_to_site

    valid = [u for u in candidates if url_belongs_to_site(u, site.home_url, site.is_domain_property)]
    rejected = len(candidates) - len(valid)

    result = url_service.add_urls(db, site, valid, source="manual", priority=priority)
    message = f"Dodano {result['added']} URL-i ({result['duplicates']} juz istnialo)."
    if rejected:
        message += f" Odrzucono {rejected} adresow spoza domeny {site.display_name}."
    flash(request, message, "success" if result["added"] else "warning")

    if submit_now and result["added"]:
        new_ids = [
            row[0]
            for row in db.execute(
                select(PageUrl.id).where(
                    PageUrl.site_id == site.id,
                    PageUrl.url.in_(valid),
                )
            ).all()
        ]
        run_in_background(f"site:{site.id}", tasks.task_submit_urls, site.id, new_ids)
        flash(request, "Zgloszenie do Google uruchomione w tle.", "success")

    return RedirectResponse(f"/sites/{site_id}?tab=urls", status_code=303)


@router.post("/action")
def bulk_action(
    request: Request,
    workspace: CurrentWorkspace,
    db: DbSession,
    action: str = Form(...),
    url_ids: Annotated[list[int], Form()] = [],  # noqa: B006
    redirect_to: str = Form("/urls"),
):
    if not url_ids:
        flash(request, "Nie zaznaczono zadnego URL-a.", "warning")
        return RedirectResponse(redirect_to, status_code=303)

    pages = list(
        db.scalars(
            select(PageUrl)
            .join(Site, Site.id == PageUrl.site_id)
            .where(PageUrl.id.in_(url_ids), Site.workspace_id == workspace.id)
        )
    )
    if not pages:
        flash(request, "Nie znaleziono wskazanych URL-i.", "error")
        return RedirectResponse(redirect_to, status_code=303)

    by_site: dict[int, list[int]] = {}
    for page in pages:
        by_site.setdefault(page.site_id, []).append(page.id)

    if action == "submit":
        for site_id, ids in by_site.items():
            run_in_background(f"site:{site_id}", tasks.task_submit_urls, site_id, ids)
        flash(request, f"Zgloszono {len(pages)} URL-i do indeksowania (w tle).", "success")
    elif action == "inspect":
        for site_id, ids in by_site.items():
            run_in_background(f"site:{site_id}", tasks.task_inspect_urls, site_id, ids)
        flash(request, f"Uruchomiono inspekcje {len(pages)} URL-i (w tle).", "success")
    elif action == "delete":
        for page in pages:
            db.delete(page)
        db.commit()
        flash(request, f"Usunieto {len(pages)} URL-i.", "success")
    elif action == "priority":
        for page in pages:
            page.priority = 10
        db.commit()
        flash(request, f"Ustawiono wysoki priorytet dla {len(pages)} URL-i.", "success")
    elif action == "reset":
        for page in pages:
            page.index_status = IndexStatus.UNKNOWN
            page.last_checked_at = None
        db.commit()
        flash(request, f"Zresetowano status {len(pages)} URL-i.", "success")
    else:
        flash(request, f"Nieznana akcja: {action}", "error")

    return RedirectResponse(redirect_to, status_code=303)


@router.get("/export.csv")
def export_csv(workspace: CurrentWorkspace, db: DbSession, site_id: int = 0, status: str = ""):
    query = (
        select(PageUrl, Site.display_name)
        .join(Site, Site.id == PageUrl.site_id)
        .where(Site.workspace_id == workspace.id)
    )
    if site_id:
        query = query.where(PageUrl.site_id == site_id)
    if status:
        query = query.where(PageUrl.index_status == status)

    buffer = io.StringIO()
    writer = csv.writer(buffer, delimiter=";")
    writer.writerow(
        [
            "strona",
            "url",
            "status",
            "coverage_state",
            "verdict",
            "ostatnia_inspekcja",
            "ostatnie_zgloszenie",
            "liczba_zgloszen",
            "zrodlo",
        ]
    )
    for page, site_name in db.execute(query).all():
        writer.writerow(
            [
                site_name,
                page.url,
                page.index_status,
                page.coverage_state or "",
                page.verdict or "",
                page.last_checked_at.isoformat() if page.last_checked_at else "",
                page.last_submitted_at.isoformat() if page.last_submitted_at else "",
                page.submit_count,
                page.source,
            ]
        )

    buffer.seek(0)
    return StreamingResponse(
        iter([buffer.getvalue()]),
        media_type="text/csv; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="indexmeplease-urls.csv"'},
    )

File: src/routes/urls.ts
import { Router } from "express";
import { sql } from "kysely";
import { db } from "../db";
import { IndexStatus, isDomainProperty } from "../db/types";
import { asyncHandler, flash, requireSite } from "../middleware/auth";
import { sha256 } from "../lib/crypto";
import { runInBackground } from "../services/scheduler";
import { listSites } from "../services/sites";
import { urlBelongsToSite } from "../services/sitemapParser";
import * as tasks from "../services/tasks";
import * as urlService from "../services/urls";
import { baseContext } from "../templating";
import { panelAuth } from "./auth";

export const urlsRouter = Router();
urlsRouter.use(...panelAuth);

const PER_PAGE = 100;

function formBool(value: unknown): boolean {
  return value === "on" || value === "true" || value === "1" || value === true;
}

urlsRouter.get(
  "/",
  asyncHandler(async (req, res) => {
    const workspace = req.workspace!;
    const siteId = Number(req.query.site_id) || 0;
    const status = typeof req.query.status === "string" ? req.query.status : "";
    const q = typeof req.query.q === "string" ? req.query.q : "";
    const page = Math.max(1, Number(req.query.page) || 1);

    let query = db
      .selectFrom("urls")
      .innerJoin("sites", "sites.id", "urls.site_id")
      .where("sites.workspace_id", "=", workspace.id);
    if (siteId) query = query.where("urls.site_id", "=", siteId);
    if (status) query = query.where("urls.index_status", "=", status as IndexStatus);
    if (q) query = query.where("urls.url", "like", `%${q}%`);

    const totalRow = await query
      .select(sql<number>`COUNT(urls.id)`.as("total"))
      .executeTakeFirst();
    const total = Number(totalRow?.total ?? 0);

    const rows = await query
      .selectAll("urls")
      .orderBy(sql`urls.last_checked_at IS NULL`, "desc")
      .orderBy("urls.id", "desc")
      .offset((page - 1) * PER_PAGE)
      .limit(PER_PAGE)
      .execute();

    const sites = await listSites(workspace.id);
    const siteMap = Object.fromEntries(sites.map((s) => [s.id, s]));

    res.render(
      "urls.html",
      baseContext(req, {
        user: req.user,
        workspace,
        rows,
        sites,
        site_map: siteMap,
        total,
        page,
        pages: Math.max(1, Math.ceil(total / PER_PAGE)),
        filter_site: siteId,
        filter_status: status,
        query: q,
        statuses: Object.values(IndexStatus),
        stats: await urlService.workspaceUrlStats(workspace.id),
        active_page: "urls",
      }),
    );
  }),
);

urlsRouter.post(
  "/add",
  asyncHandler(async (req, res) => {
    const siteId = Number(req.body.site_id);
    const site = await requireSite(req, siteId);
    let blob = String(req.body.urls_blob ?? "");

    const file = req.file;
    if (file?.buffer) {
      try {
        blob += `\n${file.buffer.toString("utf8")}`;
      } catch {
        flash(req, "Nie udalo sie odczytac pliku.", "error");
      }
    }

    const candidates = urlService.parseUrlBlob(blob);
    if (candidates.length === 0) {
      flash(req, "Nie znaleziono zadnego poprawnego adresu URL.", "error");
      return res.redirect(303, `/sites/${siteId}?tab=urls`);
    }

    const valid = candidates.filter((u) =>
      urlBelongsToSite(u, site.home_url, isDomainProperty(site)),
    );
    const rejected = candidates.length - valid.length;
    const priority = Number(req.body.priority) || 0;
    const result = await urlService.addUrls(site, valid, "manual", new Map(), priority);

    let message = `Dodano ${result.added} URL-i (${result.duplicates} juz istnialo).`;
    if (rejected) {
      message += ` Odrzucono ${rejected} adresow spoza domeny ${site.display_name}.`;
    }
    flash(req, message, result.added ? "success" : "warning");

    if (formBool(req.body.submit_now) && result.added) {
      const hashes = valid.map((u) => sha256(u));
      const newIds = (
        await db
          .selectFrom("urls")
          .select("id")
          .where("site_id", "=", site.id)
          .where("url_hash", "in", hashes)
          .execute()
      ).map((r) => r.id);
      runInBackground(`site:${site.id}`, () => tasks.taskSubmitUrls(site.id, newIds));
      flash(req, "Zgloszenie do Google uruchomione w tle.", "success");
    }

    res.redirect(303, `/sites/${siteId}?tab=urls`);
  }),
);

urlsRouter.post(
  "/action",
  asyncHandler(async (req, res) => {
    const action = String(req.body.action ?? "");
    const redirectTo = String(req.body.redirect_to ?? "/urls");
    const rawIds = req.body.url_ids;
    const urlIds = (Array.isArray(rawIds) ? rawIds : rawIds ? [rawIds] : [])
      .map((v) => Number(v))
      .filter((n) => Number.isFinite(n) && n > 0);

    if (urlIds.length === 0) {
      flash(req, "Nie zaznaczono zadnego URL-a.", "warning");
      return res.redirect(303, redirectTo);
    }

    const pages = await db
      .selectFrom("urls")
      .innerJoin("sites", "sites.id", "urls.site_id")
      .selectAll("urls")
      .where("urls.id", "in", urlIds)
      .where("sites.workspace_id", "=", req.workspace!.id)
      .execute();

    if (pages.length === 0) {
      flash(req, "Nie znaleziono wskazanych URL-i.", "error");
      return res.redirect(303, redirectTo);
    }

    const bySite = new Map<number, number[]>();
    for (const page of pages) {
      const list = bySite.get(page.site_id) ?? [];
      list.push(page.id);
      bySite.set(page.site_id, list);
    }

    if (action === "submit") {
      for (const [siteId, ids] of bySite) {
        runInBackground(`site:${siteId}`, () => tasks.taskSubmitUrls(siteId, ids));
      }
      flash(req, `Zgloszono ${pages.length} URL-i do indeksowania (w tle).`, "success");
    } else if (action === "inspect") {
      for (const [siteId, ids] of bySite) {
        runInBackground(`site:${siteId}`, () => tasks.taskInspectUrls(siteId, ids));
      }
      flash(req, `Uruchomiono inspekcje ${pages.length} URL-i (w tle).`, "success");
    } else if (action === "delete") {
      await db
        .deleteFrom("urls")
        .where(
          "id",
          "in",
          pages.map((p) => p.id),
        )
        .execute();
      flash(req, `Usunieto ${pages.length} URL-i.`, "success");
    } else if (action === "priority") {
      await db
        .updateTable("urls")
        .set({ priority: 10 })
        .where(
          "id",
          "in",
          pages.map((p) => p.id),
        )
        .execute();
      flash(req, `Ustawiono wysoki priorytet dla ${pages.length} URL-i.`, "success");
    } else if (action === "reset") {
      await db
        .updateTable("urls")
        .set({ index_status: IndexStatus.UNKNOWN, last_checked_at: null })
        .where(
          "id",
          "in",
          pages.map((p) => p.id),
        )
        .execute();
      flash(req, `Zresetowano status ${pages.length} URL-i.`, "success");
    } else {
      flash(req, `Nieznana akcja: ${action}`, "error");
    }

    res.redirect(303, redirectTo);
  }),
);

urlsRouter.get(
  "/export.csv",
  asyncHandler(async (req, res) => {
    const workspace = req.workspace!;
    const siteId = Number(req.query.site_id) || 0;
    const status = typeof req.query.status === "string" ? req.query.status : "";

    let query = db
      .selectFrom("urls")
      .innerJoin("sites", "sites.id", "urls.site_id")
      .select([
        "sites.display_name as site_name",
        "urls.url",
        "urls.index_status",
        "urls.coverage_state",
        "urls.verdict",
        "urls.last_checked_at",
        "urls.last_submitted_at",
        "urls.submit_count",
        "urls.source",
      ])
      .where("sites.workspace_id", "=", workspace.id);
    if (siteId) query = query.where("urls.site_id", "=", siteId);
    if (status) query = query.where("urls.index_status", "=", status as IndexStatus);

    const rows = await query.execute();
    const header =
      "strona;url;status;coverage_state;verdict;ostatnia_inspekcja;ostatnie_zgloszenie;liczba_zgloszen;zrodlo\n";
    const body = rows
      .map((row) =>
        [
          row.site_name,
          row.url,
          row.index_status,
          row.coverage_state || "",
          row.verdict || "",
          row.last_checked_at?.toISOString() ?? "",
          row.last_submitted_at?.toISOString() ?? "",
          row.submit_count,
          row.source,
        ]
          .map((v) => `"${String(v).replace(/"/g, '""')}"`)
          .join(";"),
      )
      .join("\n");

    res.setHeader("Content-Disposition", 'attachment; filename="indexmenow-urls.csv"');
    res.type("text/csv; charset=utf-8").send(header + body);
  }),
);
